phabricatoremails/config.py

```diff
@@ -0,0 +1,102 @@
+import smtplib
+from configparser import ConfigParser
+from dataclasses import dataclass
+from logging import Logger
+from typing import Any
+
+import psycopg2
+from phabricatoremails.logging import create_dev_logger, create_logger
+from phabricatoremails.mail import SesMail, SmtpMail, FsMail
+from phabricatoremails.progress import StaticProgress, PhabricatorProgress
+from phabricatoremails.source import FileSource, PhabricatorSource
+from psycopg2.extras import DictCursor
+
+
+def _parse_logger(config: ConfigParser):
+    if config.has_section("dev"):
+        return create_dev_logger()
+    else:
+        return create_logger()
+
+
+def _parse_source(config: ConfigParser, logger: Logger):
+    dev_source_file = config.get("dev", "file", fallback=None)
+    if dev_source_file:
+        return FileSource(dev_source_file), StaticProgress(None)
+
+    host = config.get("phabricator", "host")
+    token = config.get("phabricator", "token")
+    poll_gap_seconds = int(config.get("phabricator", "poll_gap_seconds", fallback="60"))
+    override_since_key = config.get("dev", "since_key", fallback=None)
+    override_limit = config.get("dev", "limit", fallback=None)
+    if host.endswith("/"):
+        host = host[:-1]
+
+    source = PhabricatorSource(host, token, override_limit)
+    progress = (
+        StaticProgress(override_since_key)
+        if override_since_key
+        else PhabricatorProgress(logger, poll_gap_seconds)
+    )
+    return source, progress
+
+
+@dataclass
+class Settings:
+    source: Any
+    progress: Any
+    bugzilla_host: str
+    logger: Logger
+    sentry_dsn: str
+    _config: ConfigParser
+
+    def __init__(self, config: ConfigParser):
+        logger = _parse_logger(config)
+        source, progress = _parse_source(config, logger)
+        self.source = source
+        self.progress = progress
+        self.bugzilla_host = config.get("bugzilla", "host")
+        self.logger = logger
+        self.sentry_dsn = config.get("sentry", "dsn", fallback=None)
+        self._config = config
+
+    def db(self):
+        host = self._config.get("db", "host")
+        name = self._config.get("db", "name")
+        username = self._config.get("db", "username")
+        password = self._config.get("db", "password")
+
+        return psycopg2.connect(
+            host=host,
+            dbname=name,
+            user=username,
+            password=password,
+            cursor_factory=DictCursor,
+        )
+
+    def mail(self):
+        from_address = self._config.get("email", "from_address")
+
+        if self._config.has_section("email-ses"):
+            return SesMail(from_address)
+
+        if self._config.has_section("email-smtp"):
+            host = self._config.get("email-smtp", "host")
+            send_to = self._config.get("email-smtp", "send_to", fallback=None)
+            mail_server = smtplib.SMTP(host, timeout=1)
+            return SmtpMail(mail_server, from_address, send_to, self.logger)
```

Comment:
Note that, for this one, `mail-server` depends on `host`, not all three are direct parameters for `SmtpMail`.

Reply:
I want to add some additional context here: I like to think about the two sides of the coin in which code is used by humans: at "read time" and at "write time".

I think that the [above suggestions](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r403249513) are _great_ for "write time"! You only need to modify the `keys` tuple to adjust parameters, and you don't need to repeat any code. Nice :ok_hand: 

However, this compromises "read time", since developers now need to unroll the dictionary logic that's been set up. It's do-able, and almost every developer will be able to parse it and figure out what's happening, but the technique obfuscates what's happening and makes it take longer to understand. (Also, there's an argument about how tooling/mypy, if used/IDEs might struggle to understand the data flow).

Since software is read many more times than its written, I think that err-ing on the side of read-time is a stronger choice here.